recirq/qcqmc/trial_wf.py

```diff
@@ -0,0 +1,1584 @@
+        grad: An array of gradients with respect to the one- and two-body
+            parameters. The first n_orb * (n_orb + 1) // 2 parameters correspond to
+            the one-body gradients.
+    """
+    phi = get_evolved_wf(
+        one_body_params,
+        two_body_params,
+        initial_wf,
+        gate_generators,
+        n_orb,
+        restricted=restricted,
+    )[0]
+    lam = copy.deepcopy(phi)
+    lam = lam.apply(fqe_ham)
+    cost_val = fqe.vdot(lam, phi) + e_core
+
+    # 1body
+    one_body_cluster_op = get_one_body_cluster_coef(
+        one_body_params, n_orb, restricted=restricted
+    )
+    tril = np.tril_indices(n_orb, k=-1)
+    if restricted:
+        one_body_ham = fqe.get_restricted_hamiltonian((-1j * one_body_cluster_op,))
+    else:
+        one_body_ham = fqe.get_sso_hamiltonian((-1j * one_body_cluster_op,))
+    # Apply U1b^{dag}
+    phi.time_evolve(1, one_body_ham, inplace=True)
+    lam.time_evolve(1, one_body_ham, inplace=True)
+    one_body_grad = np.zeros_like(one_body_params)
+    n_one_body_params = len(one_body_params)
+    grad_position = n_one_body_params - 1
+    for iparam in range(len(one_body_params)):
+        mu_state = copy.deepcopy(phi)
+        pidx = n_one_body_params - iparam - 1
+        pidx_spin = 0 if restricted else pidx // (n_one_body_params // 2)
+        pidx_spat = pidx if restricted else pidx - (n_one_body_params // 2) * pidx_spin
+        p, q = (tril[0][pidx_spat], tril[1][pidx_spat])
+        p += n_orb * pidx_spin
+        q += n_orb * pidx_spin
+        pre_matrix = orbital_rotation_gradient_matrix(-one_body_cluster_op, p, q)
+        assert of.is_hermitian(1j * pre_matrix)
+        if restricted:
+            fqe_quad_ham_pre = fqe.get_restricted_hamiltonian((pre_matrix,))
+        else:
+            fqe_quad_ham_pre = fqe.get_sso_hamiltonian((pre_matrix,))
+        mu_state = mu_state.apply(fqe_quad_ham_pre)
+        one_body_grad[grad_position] = 2 * fqe.vdot(lam, mu_state).real
+        grad_position -= 1
+    # Get two-body contributions
+    two_body_grad = np.zeros(len(two_body_params))
+    for pidx in reversed(range(len(gate_generators))):
+        mu = copy.deepcopy(phi)
+        mu = mu.apply(gate_generators[pidx])
+        two_body_grad[pidx] = -np.real(2 * 1j * (fqe.vdot(lam, mu)))
+        phi = phi.time_evolve(-two_body_params[pidx], gate_generators[pidx])
+        lam = lam.time_evolve(-two_body_params[pidx], gate_generators[pidx])
+
+    return cost_val, np.concatenate((two_body_grad, one_body_grad))
+
+
+def get_pp_plus_params(
+    *,
+    hamiltonian_data: HamiltonianData,
+    restricted: bool = False,
+    random_parameter_scale: float = 1.0,
+    initial_orbital_rotation: Optional[np.ndarray] = None,
+    heuristic_layers: Tuple[LayerSpec, ...],
+    do_pp: bool = True,
+    n_optimization_restarts: int = 1,
+    do_print: bool = True,
+    use_fast_gradients: bool = False,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
```

Comment:
This is also a long and complicated function with no docstring or comments.

Reply:
Split it up and added a reference to the paper.